### ISSUE TITLE
Master password kept in memory after login

### ISSUE THREAD

**Issue by ledeniz** [bitwarden/clients#6231]
### Steps To Reproduce

1. Start Bitwarden Desktop
2. Log in to your Bitwarden Vault
3. Create memory dump files via Windows task manager
4. Investigate content of the files, e.g. by running `strings`

### Expected Result

The master password **is not** present in plain text in the memory dump of any Bitwarden process.

### Actual Result

The master password **is** present in plain text in the memory dump of any Bitwarden process.

### Screenshots or Videos

![image](https://github.com/bitwarden/clients/assets/11775113/1362a517-0fe9-447f-8d07-cfa23b04fcbe)

![image](https://github.com/bitwarden/clients/assets/11775113/1cdadea0-cc69-41cb-ace0-971fbcfd5e99)

### Additional Context

I stumbled over this issue in the archived `desktop` repository and realized it's still not fixed: https://github.com/bitwarden/desktop/issues/476

### Operating System

Windows

### Operating System Version

10.0.19045 Build 19045

### Installation method

Microsoft Store

### Build Version

2023.8.2

### Issue Tracking Info

- [X] I understand that work is tracked outside of Github. A PR will be linked to this issue should one be opened to address it, but Bitwarden doesn't use fields like "assigned", "milestone", or "project" to track progress.

**Comment by ledeniz** [comment]
Thanks @bwbug for your answeraand the hint to the related issue. Does this mean it's impossible to resolve (because of the underlying browser engine)? If that is the case, please feel free to close this issue

**Comment by atjbramley** [comment]
Hi @ledeniz, 

Thank you for your report - it has been escalated for further investigation. If you have more information that can help us, please add it below. 

Thanks!

**Comment by FlakyPi** [comment]
I can confirm that the same happens when using the Firefox extension. I think this is an unacceptable security issue.

**Comment by BizClimb** [comment]
Perhaps a middle ground would be to encode the plain text password into some other form of string so it's not sitting in memory as is, it's not a bulletproof solution by any means but if there was a simple hashing or encoding on what's kept in memory, then the client itself can decode the encoded password stored in memory when it's needed? This would simply allow hiding the actual plain text vault password?

**Comment by BizClimb** [comment]
Yes, it would be an "extra step" for an attacker to go through, and a skilled attacker could certainly get it if enough work is put in, but that's what security is all about, making it as hard as possible for attackers, so I would not consider this just "security theater". 

Of course risk is relative to the perceived threat model, but there are attacks that could potentially search common strings to try and extract the full vault password. Example if my password is BobIsMyName123, and an attacker search's the memory for Bob since they might know it's my name, they could find that the full vault password is BobIsMyName123 quite quickly, where as if it's hashed and not in plain text, that would not work.

Either way, you can see CERT's advice regarding handling passwords in memory, specifically plain text storage of passwords in memory [here](https://wiki.sei.cmu.edu/confluence/display/c/MSC18-C.+Be+careful+while+handling+sensitive+data,+such+as+passwords,+in+program+code)

Regardless, I wouldn't think of this being a super critical issue that needs immediate attention or anything, since a device's memory should be secured from external threats in general, but in my opinion it certainly warrants some consideration if it should be implemented at some point to increase overall security.

**Comment by dsommers** [comment]
Regardless of potential solutions, it's just obfuscation of the secret information.  And it needs to be seen in context of various threat models.

I do agree that storing the password in plain text in memory is highly questionable, but to make it unpredictable, the code gets complicated.  And is it worth the cost, when comparing it against threat models?  How likely is it this situation could occur? What kind of attack vectors would be needed to used - in addition to just do a memory dump?  Who can do trigger a memory dump?  Who else can read memory regions where this information is stored?

Potential solutions could be to create a random encryption key at program start, ensuring that key is **not** ending up in the same memory location each time and encrypt/decrypt secrets like the master password using that key.  And if that encrypted key is stored on different locations each time, it gets further more challenging.

But at some point, that encrypted information need to be decrypted and stored in plain text in memory for a shorter time ...  So while it may be possible to reduce the chances to extract the plain text password, is it worth the effort with the additional code complexity?

And also consider that additional code complexity can also introduce new attack vectors in the program, triggering this "obfuscation logic" to be more predictable - leading to the password leak.

So to say it bluntly: Who and/or what are you trying to protect this access to the master password against?

I don't believe this kind of challenge can be fully be solved in software alone to be closer to bullet perfect.  Once operating systems provides a possibility to have separate encrypted memory regions per application and the CPU is the instance orchestrating the encryption keys on behalf of the applications (ie. the OS or the application itself does not have access to the encryption key itself) - that's where we can get closer to a real solution.

Anything else, will essentially just be obfuscation of the memory content - in one or another way.

**Comment by tribut** [comment]
I've tried to reproduce this using the Firefox addon (on Linux) and the good news is that the master password is not kept in memory in the clear, neither in unlocked nor in locked state.

However, I did notice that multiple versions of the master password with typos I had made while trying to unlock the vault were still in memory. Is there some sort of special treatment of the master password that is applied only when the vault was successfully unlocked? If yes, this should also apply to rejected master passwords.

**Comment by zommuter** [comment]
Ultimately I think the only "real" solution to this and #1516  is keeping any secrets away from memory entirely, which means letting dedicated hardware handle the decryption or even better signing passkey requests only, but that would render the whole cloud service mostly moot...

**Comment by dadatuputi** [comment]
@zommuter That's my conclusion. But it seems like there's zero support for it anywhere, Military smart cards do encryption/decryption on the little chip. This is has been the standard for over a decade for DoD email. There's plenty of hardware options now we can leverage, e.g., Nitrokey HSM 2, YubiKey with PIV. I guess it just takes PKCS#11 support.

**Comment by trmartin4** [comment]
We have attempted to reproduce this internally using the latest client versions and have not been able to do so.  The only scenario that we tested that would produce this behavior is if the user is re-using their master password for another account which is stored in the vault, or if the user is storing their master password in the vault.  In those cases, the master password would appear in a memory dump as the vault contents are available in memory when the vault is unlocked.

However, if you are still able to reproduce this please feel free to re-open or create another issue with your reproduction steps so that we can be sure to address it.
